### PR TITLE
Events refactor And Tests

### PR DESCRIPTION
**Pull request by blinky3713**
This PR is a combination of #122  and #123 and replaces both of them.

### CODE REVIEW

File: src/Network/Ethereum/Web3/Contract/Events.purs [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+         -- if the original filter goes to Pending or Latest, consume as many as possible up to the chain head
+         Pending -> continueTo chainHead
+         Latest -> continueTo chainHead
+         -- if the original fitler ends at a specific block, consume as many as possible up to that block
+         -- or terminate if we're already past it
+         BN targetEnd -> 
+           if currentState.currentBlock <= targetEnd
+             then continueTo targetEnd
+             else pure currentState
+         -- otherwsie we're in Earliest, which is a degenerate case
```
Comment (safareli) [inline src/Network/Ethereum/Web3/Contract/Events.purs]: typo `otherwsie`

File: src/Network/Ethereum/Web3/Contract/Events.purs [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    -- if the chain head is less than the current block we want to process
+    -- then wait until the chain progresses
+    if chainHead < currentState.currentBlock
+       then waitForMoreBlocks
+       -- otherwise try make progress
+       else case currentState.initialFilter ^. _toBlock of
+         -- if the original filter goes to Pending or Latest, consume as many as possible up to the chain head
+         Pending -> continueTo chainHead
+         Latest -> continueTo chainHead
+         -- if the original fitler ends at a specific block, consume as many as possible up to that block
```
Comment (safareli) [inline src/Network/Ethereum/Web3/Contract/Events.purs]: typo `fitler`

File: test/Web3Spec/Contract.purs [hunk 3]
```diff
@@ -14,7 +14,7 @@ import Data.Lens ((.~))
 import Data.Maybe (Maybe(..), fromJust)
 import Data.Newtype (class Newtype, wrap)
 import Data.Symbol (SProxy)
-import Network.Ethereum.Web3 (class EventFilter, Address, ChainCursor(..), EventAction(..), HexString, Web3, Web3Error, _address, _fromBlock, _toBlock, _topics, defaultFilter, defaultTransactionOptions, embed, event, eventFilter, forkWeb3', mkAddress, mkHexString, sendTx)
```
Comment (safareli) [inline test/Web3Spec/Contract.purs]: I guess this file should be moved to
`test/web3/Web3Spec/Contract.purs`?

**Comment by safareli** [comment]
It would be great if you add a readme to chanterelle folder describing it's purpose and how to use it

File: test/Web3Spec/Contract.purs (hunk 3 above)
Comment (blinky3713) [reply test/Web3Spec/Contract.purs]: Actually I mean to delete it

**Pull request merged**